### ISSUE TITLE
feature: add support for environment specific addons

### ISSUE THREAD

**Issue by noahjahn** [aws/copilot-cli#1975]
It would be awesome to be able to have AWS resources created for services to share within a single environment.

### Use case
I have two services running as part of each environment in my copilot application. Both services required a PostgreSQL database. Instead of creating a RDS instance for each service, I would like to have a single RDS instance for the environment that either of the two services could reference different databases in.

**Comment by sharepointoscar** [comment]
I might be reading your feature request incorrectly, but doesn't this satisfy your need to deploy additional resources? https://aws.github.io/copilot-cli/docs/developing/additional-aws-resources @noahjahn

**Comment by noahjahn** [comment]
> I might be reading your feature request incorrectly, but doesn't this satisfy your need to deploy additional resources? https://aws.github.io/copilot-cli/docs/developing/additional-aws-resources @noahjahn

I think this could work, I just thought it'd be a cool to not tie the add on to a specific service but rather at a higher level.

**Comment by efekarakus** [comment]
Hi @noahjahn ! Not quite a solution but this comment: https://github.com/aws/copilot-cli/issues/1368#issuecomment-692316027 might be of interest for you.

The idea is to create a resource tied to a particular service in the addons stack, and then "export" any interesting values as SSM parameters so that they can be injected as environment variables in other services. While it's not ideal, it's a work around for now.

**Comment by toshitanian** [comment]
@efekarakus I struggled about the "environment level" storage source. As you know, service level storage add-ons like Aurora DB is deleted after the service is deleted. We made a custom cf template based on the ones you've been mentioning in the issues.
However, still it's hard to manage the environment because some error occurs while manipulating copilot environment because there's non-copilot-managed resources.
So, I want to request that the 'environment level' storage resources are supported in copilot CLI

**Comment by efekarakus** [comment]
Hi @toshitanian 👋 

That makes sense! Can you describe your architecture for me so that we can ensure we're building the right thing?

Do you have multiple services accessing the same aurora DB? or is your goal to ensure that the database is retained even if a service is removed?

**Comment by toshitanian** [comment]
@efekarakus Thank you for the quick reply!  

>Do you have multiple services accessing the same aurora DB? or is your goal to ensure that the database is retained even if a service is removed?

In our use case, we have one http service and one hourly job. Both access to the database to handle the user data. For example, we are making a survey system that users answers on the web, and hourly job summarize the survey result and send the summary to the administrator. Please see the architecture of our use case in the diagram below.

## In-depth situation
( Please look at the diagrams as well )

Firstly I created a database with `copilot storage` but found that it belongs to the "service" and hard to manage. In case I execute `copilot service delete`, the database is deleted as well. I wanted the database to be persistent because I planed to use the database from the job as well.
In addition to that, I wanted "service" not to have state so that we can try-and-error anytime.

As a workaround, I made a custom cf template of the database (please note that I made RDS instead of Aurora) using the exported variables of CloudFormation stack. It uses VPC and security groups that `copilot env` manages. But with this workaround, we have to add the security group the database allows to the manifest file, `network.vpc.security_groups`.
Also, It's hard to manage because the stacks are difference and `copilot env delete` will make error because the database uses the resource of copilot environment.

So, my proposal is to make an environment-level storage that copilot manages and take care of security group, etc.

![Untitled presentation (2)](https://user-images.githubusercontent.com/1647045/134006698-dbc02ff5-98ef-41c3-844d-a6d9d604a7fc.png)

![Untitled presentation (1)](https://user-images.githubusercontent.com/1647045/134006677-b411cf91-8e97-4f21-9b0b-dbe8a68f334a.png)

![Untitled presentation](https://user-images.githubusercontent.com/1647045/134006657-77dfadcd-7e56-42b5-91bd-7c2aef1d9cb1.png)

**Comment by efekarakus** [comment]
@toshitanian wow thank you for the awesome explanation ! The diagrams are extremely helpful! 😍 

If you don't mind I have a few alternatives proposals too until we have environment addons supported in the CLI.

### Option 1: Keeping the database part of `http-service`

One possibility is sticking with the first diagram "default add-on storage" but instead of the `hourly-job` directly querying the database, it can make HTTP requests to the `http-service`.   
For example, using [service discovery](https://aws.github.io/copilot-cli/docs/developing/service-discovery/) the `hourly-job` can make a: 
```
GET http://hourly-job.${COPILOT_SERVICE_DISCOVERY_ENDPOINT}/getHourlySummary
``` 
request and then process the response.

The database can be retained on service deletion by modifying the CloudFormation template with:
```yaml
DBCluster:
  "DeletionPolicy": "Retain"
```

This has the benefit of ensuring there is only one microservice that accesses the database.

### Option 2: Faking environment addons

One other possibility is to create a dummy Backend Service, set the `count: 0`, and create the rds addon for that service.

You can modify the security group for the cluster to use the EnvironmentSecurityGroup: 
https://github.com/aws/copilot-cli/blob/9055f4ff4fa8bc0d0f2ec48bf7c8624c990b0c48/internal/pkg/template/templates/environment/cf.yml#L324-L327
So that all the tasks in the environment can access the database.

This should allow `copilot svc delete` and `copilot env delete` to be work properly.

---------

Thanks for the feature request, the description makes sense!

**Comment by toshitanian** [comment]
@efekarakus Thank you for the in-depth descriptions! Will try out the both for evaluation and will change our architecture 👍

**Comment by efekarakus** [comment]
Hi folks! @Lou1415926 posted a proposal to support env addons 🚀 https://github.com/aws/copilot-cli/issues/4219 we would love to hear your feedback!

**Comment by Lou1415926** [comment]
Hello all! The environment addons feature is now released in [v1.25.0](https://github.com/aws/copilot-cli/releases/tag/v1.25.0) 🎉🚀!
